### ISSUE TITLE
split TypeScript version into its own module

### ISSUE THREAD

**Issue by bcoe** [demurgos/v8-coverage#13]
Hey @demurgos 👋 would you be open to pulling the TypeScript version out into its own GitHub repository and library on npm? Perhaps you could call it `v8-coverage-merge` which is closer to what It's being used for in c8, and now experimentally in Jest?

In practice I've found that this library works great for my need of merging coverage reports, but bit rot a bit (It's hard to get the TypeScript to compile or the Gulp tasks to run).

Any ways, I'd rather get back on the mainline dependency, and would rather it be its own stand-alone library.

Thanks for the hard work.

CC: @SimenB

**Comment by demurgos** [comment]
Hi,
Okay, I'll try to split it into a smaller repo that only handles merging.

**Comment by SimenB** [comment]
Related, you might consider getting the v8 types from the node types on definetely typed

https://github.com/demurgos/v8-coverage/blob/master/ts/src/lib/types.ts is essentially https://github.com/DefinitelyTyped/DefinitelyTyped/blob/1d3c2241784d39d9378c99473f629cb77f5c8e89/types/node/inspector.d.ts#L1631-L1643

**Comment by demurgos** [comment]
I had some time to look more into the repo.
The majority of the repo is already required for the TS part: source, documentation, test and benchmark files. The two extra directories are `rs` for the Rust implementation and `node` for the native `.node` version (using the Rust code). The `.node` implementation is 2-3 times faster on my computer: since merging is a CPU-intensive task it's a nice boost, but it is also slightly out of date (it does not handle mixed precise and imprecise coverage).

Regarding [the code itself](https://github.com/demurgos/v8-coverage/tree/master/ts/src/lib), most of it is required to for the merge algorithm. The only module that is "apart" is `ascii.ts`, it's more an internal module used to read the ASCII diagrams describing the test samples. I included it in the main package (and not just the test files) because it could be used to dump a `FunctionCoverage` to visualize it without all the higher-level rendering of `c8`.

Using the `node/inspector` types is a good idea. I still think that the types used by this lib should be redefined here because of two reasons: this library uses an extension of the V8 format to merge precise and imprecise coverage without information loss, and it avoids platform specific dependencies (the library doesn't have any dependency). Typescript should allow the types produced by the inspector to match the types expected by the library (in both directions), and catch eventual incompatibilities if the V8 format ever changes. The `node/inspector` should still be used in the tests to check that both declarations are compatible (having it in `devDependencies` is OK).
If the inspector types were available as a standalone dependency (not part of Node's types) I could add them as a regular dependency with an `^x.y.z` range. Node's types use ambient declarations that are more akin to environment declarations like `dom.ts`: you _really_ want a single version of these declarations per project so it does not play well with npm's model that allows duplicated versions of regular dependencies.

I'd like to work on a `1.0.0` release of the library. Unless the V8 format changes, the library is already working fine. I had less time for code coverage tools in general, but I should be able to finish it during december. I would still like to provide `.node` and `.wasm` version of the library because of the perf benefits and because I am more confident in the reliability of the lib (especially the merge part) by having two implementation. I'll also make sure that you can develop the library without exposing `gulp`. I'd still prefer to use it internally: it allows me to share my build config between all my projects, but it shouldn't be required for regular development.

@bcoe I tracked the changes on your fork. There's only [one change between this version and your fork](https://github.com/bcoe/v8-coverage/pull/1/files#diff-9fb408071ab0fd3e460fc0afdeecf774L26), do you remember why you had to remove normalization when there's a single process coverage? Normalization is there to ensure determinism of the `merge` function.

**Comment by bcoe** [comment]
@demurgos yes, even when there was a single report it's possible that the same script occurs multiple times; an example of this in action is clearing the require cache.

The logic:

```js
if (processCovs.length === 1) {
    const merged: ProcessCov = processCovs[0];
    deepNormalizeProcessCov(merged);
    return merged;
  }
```

☝️ did not seem to take this into account appropriately, and lead to corrupted report output by the time Istanbul reports were generated -- I forget the exact symptom, it was either jumbled coverage reports, or coverage reports showing up multiple times.

Allowing the same code path to execute on a single report, vs., multiple reports addressed the problem, and did not seem to be a significant performance hit in large codebases like Node.js -- I'd advocate executing the same code-path for a single report as multiple reports, rather than creating an edge-case.

**Comment by demurgos** [comment]
@bcoe 
Thanks for the info. It may be related to precise and non-precise coverage. I'll look more into it.

Also replying to [https://github.com/facebook/jest/pull/8596#discussion_r355195831](this comment):
> Perhaps we could move this to "Istanbul/merge-v8-coverage" or "merge-v8-coverage", @demurgos do you have a preference? Would you like to work on this library within the Istanbul organization?

Yes, it would make sense to move it to the istanbul org so there are more maintainers.

**Comment by demurgos** [comment]
Even without mixing precise and imprecise coverage, I think it may occur when playing with Node's module cache: load a module, clear the cache, load the module again.

**Comment by SimenB** [comment]
@bcoe @demurgos any news here? Booth Jest and C8 still uses the fork

**Comment by SimenB** [comment]
@bcoe @demurgos almost a year later, any news? 😀

**Comment by demurgos** [comment]
Hi, sorry for the long delay.
I am actually in the process of updating the libraries I maintain: I will make sure to clean-up this repo, fix this issue and release a 1.0 version.

**Comment by SimenB** [comment]
Woah, quick answer! Awesome, thank you!
